src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs

```diff
@@ -3,16 +3,214 @@
+        /// <returns>A task that represents the entire for each operation.</returns>
+        /// <remarks>The operation will execute at most <see cref="Environment.ProcessorCount"/> operations in parallel.</remarks>
+        public static Task ForAsync<T>(T fromInclusive, T toExclusive, CancellationToken cancellationToken, Func<T, CancellationToken, ValueTask> body)
+            where T : notnull, IBinaryInteger<T>
+        {
+            if (fromInclusive is null) throw new ArgumentNullException(nameof(fromInclusive));
+            if (toExclusive is null) throw new ArgumentNullException(nameof(toExclusive));
+            ArgumentNullException.ThrowIfNull(body);
+
+            return ForAsync(fromInclusive, toExclusive, DefaultDegreeOfParallelism, TaskScheduler.Default, cancellationToken, body);
+        }
+
+        /// <summary>Executes a for loop in which iterations may run in parallel.</summary>
+        /// <param name="fromInclusive">The start index, inclusive.</param>
+        /// <param name="toExclusive">The end index, exclusive.</param>
+        /// <param name="parallelOptions">An object that configures the behavior of this operation.</param>
+        /// <param name="body">An asynchronous delegate that is invoked once per element in the data source.</param>
+        /// <exception cref="ArgumentNullException">The exception that is thrown when the <paramref name="body"/> argument is null.</exception>
+        /// <returns>A task that represents the entire for each operation.</returns>
+        /// <remarks>The operation will execute at most <see cref="Environment.ProcessorCount"/> operations in parallel.</remarks>
+        public static Task ForAsync<T>(T fromInclusive, T toExclusive, ParallelOptions parallelOptions, Func<T, CancellationToken, ValueTask> body)
+            where T : notnull, IBinaryInteger<T>
+        {
+            if (fromInclusive is null) throw new ArgumentNullException(nameof(fromInclusive));
+            if (toExclusive is null) throw new ArgumentNullException(nameof(toExclusive));
+            ArgumentNullException.ThrowIfNull(parallelOptions);
+            ArgumentNullException.ThrowIfNull(body);
+
+            return ForAsync(fromInclusive, toExclusive, parallelOptions.EffectiveMaxConcurrencyLevel, parallelOptions.EffectiveTaskScheduler, parallelOptions.CancellationToken, body);
+        }
+
+        /// <summary>Executes a for each operation on an <see cref="IEnumerable{TSource}"/> in which iterations may run in parallel.</summary>
+        /// <typeparam name="T">The type of the data in the source.</typeparam>
+        /// <param name="fromInclusive">The start index, inclusive.</param>
+        /// <param name="toExclusive">The end index, exclusive.</param>
+        /// <param name="dop">A integer indicating how many operations to allow to run in parallel.</param>
```

Comment:
Optional since this is private.

Took me a second to realize `dop` was referring to "degree of parallelism".

```suggestion
        /// <param name="dop">The degree of parallelism, or the number of operations to allow to run in parallel.</param>
```